mingw.invoke.h

```diff
@@ -21,27 +21,35 @@ namespace mingw_stdthread
 {
 namespace detail
 {
-//  For compatibility, implement std::invoke for C++11 and C++14
-#if __cplusplus < 201703L
   template<bool PMemFunc, bool PMemData>
   struct Invoker
   {
     template<class F, class... Args>
+#if __cplusplus >= 201703L
+    inline static typename std::invoke_result<F, Args...>::type invoke (F&& f, Args&&... args)
+#else
     inline static typename std::result_of<F(Args...)>::type invoke (F&& f, Args&&... args)
+#endif
     {
+#if __cplusplus >= 201703L
+      return std::__invoke(std::forward<F>(f), std::forward<Args>(args)...);
```

Comment:
It's better not to rely on the current library implementation, and instead rely on [the standard](https://en.cppreference.com/w/cpp/utility/functional/invoke).

Reply:
On the standart std::result_of is removed in C++20.
The latest htj2k codecs are already C++20.

Reply:
I was thinking about Invoke a callable object in gcc 12.0.0
```
#if __cplusplus >= 201703L
# define __cpp_lib_invoke 201411L
# if __cplusplus > 201703L
#  define __cpp_lib_constexpr_functional 201907L
# endif

  /// Invoke a callable object.
  template<typename _Callable, typename... _Args>
    inline _GLIBCXX20_CONSTEXPR invoke_result_t<_Callable, _Args...>
    invoke(_Callable&& __fn, _Args&&... __args)
    noexcept(is_nothrow_invocable_v<_Callable, _Args...>)
    {
      return std::__invoke(std::forward<_Callable>(__fn),
			   std::forward<_Args>(__args)...);
    }

#if __cplusplus > 202002L
# define __cpp_lib_invoke_r 202106L

  /// Invoke a callable object and convert the result to _Res.
  template<typename _Res, typename _Callable, typename... _Args>
    requires is_invocable_r_v<_Res, _Callable, _Args...>
    constexpr _Res
    invoke_r(_Callable&& __fn, _Args&&... __args)
    noexcept(is_nothrow_invocable_r_v<_Res, _Callable, _Args...>)
    {
      return std::__invoke_r<_Res>(std::forward<_Callable>(__fn),
				   std::forward<_Args>(__args)...);
    }
#endif // C++23
#endif // C++17
 ```

Reply:
This is already handled in the existing library code. Are you seeing errors during compilation?

If you had not removed the line `#if __cplusplus < 201703L` and its corresponding lines
```
#else
using std::invoke;
#endif
```
Then the compatibility implementation would not be selected when compiling in C++20 (and, in fact, would not be compiled at all). Instead, the line `using std::invoke;` would pull in the standard library's `invoke` implementation, which then uses `invoke_result` to return type.

In short, you intend to add C++20 compatibility to code that is `#ifdef`'ed away for C++17 and higher.

Reply:
I am an amateur of programming. So out of boredom he asks questions.